### PR TITLE
feat: intake

### PR DESCRIPTION
**Pull request by minhd-vu**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -49,4 +49,32 @@ public static class OperatorConstants {
     public static final double RIGHT_X_DEADBAND = 0.1;
     public static final double TURN_CONSTANT = 6;
   }
+
+  public static class ManipulatorConstants {
```
Comment (minhd-vu) [inline src/main/java/frc/robot/Constants.java]:
```suggestion
  public static class IntakeConstants {
```

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -49,4 +49,32 @@ public static class OperatorConstants {
...
+    public static final double SHOOTER_kFF = 0.000082;
+    public static final double SHOOTER_kP = 0.00045;
+    public static final double OUTTAKE_SPEED = -0.5;
+    public static final double INTAKE_SPEED = 0.75;
+    public static final double INTAKE_REST_SPEED = 0.00;
+    public static final double SHOOTER_THRESHOLD = 150;
+    public static final double PUKE_SPEED = -1;
+    public static final double TRAP_SPEED = 1900;
+    public static final double LOADBACK_SPEED = -0.08; // MAYBE BACK AND FORTH WAS THE MOVE
+    public static final double LOADING_SPEED = 1; // (Note fixed its deformity)
```
Comment (minhd-vu) [inline src/main/java/frc/robot/Constants.java]: I don' think we use a lot of these constants so remove the ones we don't use.

File: src/main/java/frc/robot/subsystems/ManipulatorSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.revrobotics.CANSparkFlex;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.PIDSubsystem;
+import edu.wpi.first.wpilibj2.command.PIDSubsystem;
+import frc.robot.Constants;
+import frc.robot.Constants.ManipulatorConstants;
+
+public class ManipulatorSubsystem extends PIDSubsystem {
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/ManipulatorSubsystem.java]:
```suggestion
public class IntakeSubsystem extends PIDSubsystem {
```

Comment (minhd-vu) [reply src/main/java/frc/robot/subsystems/ManipulatorSubsystem.java]: We don't need this to be a PIDSubsystem, it's very basic and should just do voltage intake.

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -49,4 +56,13 @@ public static class OperatorConstants {
     public static final double RIGHT_X_DEADBAND = 0.1;
     public static final double TURN_CONSTANT = 6;
   }
+
+  public static class IntakeConstants {
+    public static final int SHOOTER_MOTOR = 19;
```
Comment (minhd-vu) [inline src/main/java/frc/robot/Constants.java]:
```suggestion
    public static final int INTAKE_MOTOR = 19;
```

File: src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -49,4 +56,13 @@ public static class OperatorConstants {
...
+    public static final Measure<Current>  MOTOR_CURRENT_LIMIT = Amp.of(40);
+    public static final Measure<Velocity<Angle>> OUTTAKE_SPEED = RPM.of(-0.5);
+    public static final Measure<Velocity<Angle>> INTAKE_SPEED = RPM.of(0.75);
+    public static final Measure<Velocity<Angle>> INTAKE_REST_SPEED = RPM.of(0);
+    public static final double SHOOTER_THRESHOLD = 150;
```
Comment (minhd-vu) [inline src/main/java/frc/robot/Constants.java]:
```suggestion
```
We don't use this.

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    intakeMotor.burnFlash();
+  }
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
+    super.periodic();
+  }
+
+  public double getIntakeVelocity() {
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
  public double getVelocity() {
```

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    return intakeMotor.getEncoder().getVelocity();
+  }
+
+  public void set(Measure<Velocity<Angle>> speed) {
+    intakeMotor.set(-speed.in(RPM));
+  }
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]: This is actually a value from -1 to 1 so we don't need a measure here.

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    public Command intakeBucket(IntakeSubsystem intake, Measure<Velocity<Angle>> intakeSpeed) {
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]: Same here.

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import com.revrobotics.CANSparkFlex;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class IntakeSubsystem extends SubsystemBase {
+  /** Creates a new ShooterSubsystem. */
+  CANSparkFlex intakeMotor;
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
  private final CANSparkFlex intakeMotor = new CANSparkFlex(Constants.IntakeConstants.INTAKE_MOTOR, MotorType.kBrushless);
```

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.revrobotics.CANSparkFlex;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class IntakeSubsystem extends SubsystemBase {
+  /** Creates a new ShooterSubsystem. */
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
```

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  CANSparkFlex intakeMotor;
+
+  public IntakeSubsystem() {
+    intakeMotor = new CANSparkFlex(Constants.IntakeConstants.INTAKE_MOTOR, MotorType.kBrushless);
+    // shooterMotor.restoreFactoryDefaults();
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
```

File: src/main/java/frc/robot/Constants.java [hunk 12]
```diff
@@ -4,9 +4,13 @@
 
 package frc.robot;
 
+import static edu.wpi.first.units.Units.Amp;
```
Comment (minhd-vu) [inline src/main/java/frc/robot/Constants.java]:
```suggestion
import static edu.wpi.first.units.Units.*;
```

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class IntakeSubsystem extends SubsystemBase {
+  /** Creates a new ShooterSubsystem. */
+  CANSparkFlex intakeMotor;
+
+  public IntakeSubsystem() {
+    intakeMotor = new CANSparkFlex(Constants.IntakeConstants.INTAKE_MOTOR, MotorType.kBrushless);
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
```

**Comment by RealSirLancelot** [comment]
```﻿﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿﻿ edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364): Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ Loop time of 0.02s overrun
 ﻿﻿ edu.wpi.first.wpilibj.IterativeRobotBase.printLoopOverrunMessage(IterativeRobotBase.java:412) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.IterativeRobotBase.printLoopOverrunMessage(IterativeRobotBase.java:412): Loop time of 0.02s overrun ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ 	IntakeSubsystem.periodic(): 0.000019s
	SwerveSubsystem.periodic(): 0.000018s
	buttons.run(): 0.000111s
 ﻿﻿ edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62): 	IntakeSubsystem.periodic(): 0.000019s ﻿
﻿﻿﻿﻿﻿﻿ 	SwerveSubsystem.periodic(): 0.000018s ﻿
﻿﻿﻿﻿﻿﻿ 	buttons.run(): 0.000111s ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ 	SmartDashboard.updateValues(): 0.000798s
	robotPeriodic(): 0.006474s
	LiveWindow.updateValues(): 0.000000s
	Shuffleboard.update(): 0.000018s
	disabledPeriodic(): 0.038844s
 ﻿﻿ edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62): 	SmartDashboard.updateValues(): 0.000798s ﻿
﻿﻿﻿﻿﻿﻿ 	robotPeriodic(): 0.006474s ﻿
﻿﻿﻿﻿﻿﻿ 	LiveWindow.updateValues(): 0.000000s ﻿
﻿﻿﻿﻿﻿﻿ 	Shuffleboard.update(): 0.000018s ﻿
﻿﻿﻿﻿﻿﻿ 	disabledPeriodic(): 0.038844s ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿﻿ edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364): Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ Loop time of 0.02s overrun
 ﻿﻿ edu.wpi.first.wpilibj.IterativeRobotBase.printLoopOverrunMessage(IterativeRobotBase.java:412) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.IterativeRobotBase.printLoopOverrunMessage(IterativeRobotBase.java:412): Loop time of 0.02s overrun ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿﻿ edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.DriverStation.reportJoystickUnplugged﻿Warning﻿(DriverStation.java:1364): Joystick Button 1 on port 0 not available, check if controller is plugged in ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ 	IntakeSubsystem.periodic(): 0.000056s
	SwerveSubsystem.periodic(): 0.000037s
	buttons.run(): 0.000964s
 ﻿﻿ edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62): 	IntakeSubsystem.periodic(): 0.000056s ﻿
﻿﻿﻿﻿﻿﻿ 	SwerveSubsystem.periodic(): 0.000037s ﻿
﻿﻿﻿﻿﻿﻿ 	buttons.run(): 0.000964s ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 1 ﻿﻿ 	SmartDashboard.updateValues(): 0.000463s
	robotPeriodic(): 0.008014s
	LiveWindow.updateValues(): 0.000019s
	Shuffleboard.update(): 0.000000s
	disabledPeriodic(): 0.039307s
 ﻿﻿ edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62) ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ ﻿Warning﻿ at edu.wpi.first.wpilibj.Tracer.lambda$printEpochs$0(Tracer.java:62): 	SmartDashboard.updateValues(): 0.000463s ﻿
﻿﻿﻿﻿﻿﻿ 	robotPeriodic(): 0.008014s ﻿
﻿﻿﻿﻿﻿﻿ 	LiveWindow.updateValues(): 0.000019s ﻿
﻿﻿﻿﻿﻿﻿ 	Shuffleboard.update(): 0.000000s ﻿
﻿﻿﻿﻿﻿﻿ 	disabledPeriodic(): 0.039307s ﻿
﻿﻿﻿﻿﻿﻿  ﻿
﻿﻿﻿﻿﻿﻿ CommandScheduler loop overrun ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 44004 ﻿﻿ FRC: The Driver Station has lost communication with the robot. ﻿﻿ Driver Station ﻿﻿﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 0 ﻿﻿  ﻿﻿  ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ 0 files downloaded to C:\Users\Public\Documents\FRC\Log Files\WPILogs ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 0 ﻿﻿  ﻿﻿  ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ 0 files downloaded to C:\Users\Public\Documents\FRC\Log Files\WPILogs ﻿
﻿﻿﻿﻿﻿﻿﻿﻿Warning ﻿﻿ 0 ﻿﻿  ﻿﻿  ﻿﻿﻿
﻿﻿﻿﻿﻿﻿ 0 files downloaded to C:\Users\Public\Documents\FRC\Log Files\WPILogs ﻿
```

File: src/main/java/frc/robot/RobotContainer.java [hunk 14]
```diff
@@ -81,6 +83,12 @@ private void configureBindings() {
     driverXbox.start().whileTrue(Commands.none());
     driverXbox.back().whileTrue(Commands.none());
     driverXbox.leftBumper().whileTrue(Commands.runOnce(drivebase::lock, drivebase).repeatedly());
+    driverXbox
+        .leftTrigger()
+        .whileTrue(intake.intakeBucket(intake, Constants.IntakeConstants.INTAKE_SPEED,true));
```
Comment (minhd-vu) [inline src/main/java/frc/robot/RobotContainer.java]:
```suggestion
        .whileTrue(intake.intakeBucket(intake, Constants.IntakeConstants.INTAKE_SPEED, true));
```

File: src/main/java/frc/robot/RobotContainer.java [hunk 15]
```diff
@@ -81,6 +83,12 @@ private void configureBindings() {
...
+    driverXbox
+        .rightTrigger()
+        .whileTrue(intake.intakeBucket(intake, Constants.IntakeConstants.OUTTAKE_SPEED,false));
```
Comment (minhd-vu) [inline src/main/java/frc/robot/RobotContainer.java]:
```suggestion
        .whileTrue(intake.intakeBucket(intake, Constants.IntakeConstants.OUTTAKE_SPEED, false));
```

File: src/main/java/frc/robot/subsystems/IntakeSubsystem.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
+    super.periodic();
+  }
+
+  public double getVelocity() {
+    return intakeMotor.getEncoder().getVelocity();
+  }
```
Comment (minhd-vu) [inline src/main/java/frc/robot/subsystems/IntakeSubsystem.java]:
```suggestion
```

**Comment by minhd-vu** [comment]
Just fix merge conflicts and should be good to go

**Review by RealSirLancelot — APPROVED** [review]

**Pull request merged**